### DISCUSSION TITLE
PorousFlow: EOS for Water, Salt, Gas with precipitation and dissolution at elevated temperatures

### DISCUSSION THREAD

**Discussion by batodon** [idaholab/moose#23701] [Q&A Modules: General] [answered]
Hello, this issue is related to ticket #23280. It turns out that the optimal EOS for my simulation should have salt precipitation and dissolution and should also account for porosity and permeability variation. TOUGH2’s EWASG (Battistelli,1997), instead of EOS4, seems the most appropriate. I, however, struggle to get the EWASG equivalent in PF. The closest I could get are `PorousFlowWaterNCG ` and `PorousFlowBrineCO2`. The latter seems more suitable, so I tried simulating TOUGH2’s EWASG example with it, but my simulation did not converge. It is probably because the `PorousFlowBrineCO2` does not include the effect of dissolution and precipitation of halite. Could you please recommend a PF fluid state module that can do this? Or do I need to make any code modifications to `PorousFlowBrineCO2`? I attached my input file for the EWASG example that I tried. Thanks.

```
[Mesh]
[gen]
  type = GeneratedMeshGenerator
  dim = 2
  xmin = 5
  xmax = 10000 #10000
  nx = 100        #100    
  bias_x = 1.01
  ymin = -500
  ymax = 0.0
  ny = 1
  []
  coord_type = RZ
[]


[FluidProperties]
  [co2]
    type = CO2FluidProperties
  []
  [brine]
    type = BrineFluidProperties
  []
  [true_water]
    type = Water97FluidProperties
  []
  [tabulated_water]
    type = TabulatedBicubicFluidProperties
    fp = true_water
    temperature_max = 900
    temperature_min = 250
    pressure_max = 1E8
    interpolated_properties = 'density viscosity enthalpy internal_energy'
    fluid_property_file = water97_tabulated_11.csv
  []
  [true_co2]
    type = CO2FluidProperties
  []
  [tabulated_co2]
    type = TabulatedBicubicFluidProperties
    fp = true_co2
    temperature_max = 900
    temperature_min = 250
    pressure_max = 1E8
    interpolated_properties = 'density viscosity enthalpy internal_energy'
    fluid_property_file = co2_tabulated_11.csv
  []
[]


[UserObjects]
  [dictator]
    type = PorousFlowDictator
    porous_flow_vars = 'pgas zi temperature'
    number_fluid_phases = 2
    number_fluid_components = 3
  []
  [pc]
    type = PorousFlowCapillaryPressureConst
    pc = 0
  []
  [fs]
    type = PorousFlowBrineCO2
    brine_fp = brine
   co2_fp = co2
#    type = PorousFlowWaterNCG
#    water_fp = true_water
#    gas_fp =  true_co2           
    capillary_pressure = pc
  []
[]

[GlobalParams]
  PorousFlowDictator = dictator
  gravity = '0.8  0.8  0.8'
  temperature_unit = Celsius
[]

[Variables]
  [pgas]
    initial_condition = 6e6
  []
  [temperature]
    initial_condition = 275.55
  []
  [zi]
  []
[]

[ICs]
  [zi]
    type = PorousFlowFluidStateIC
    variable = zi
    saturation = 0.45
    gas_porepressure = pgas
    temperature = temperature
    fluid_state = fs
  []
[]


[AuxVariables]
  [Xnacl]
    initial_condition = 0.3
  []
 [sgas]
   order = CONSTANT
   family = MONOMIAL
 []
  [sbrine]
   order = CONSTANT
   family = MONOMIAL
  []
[]


[AuxKernels]
  [sgas]
    type = PorousFlowPropertyAux
    property = saturation
    phase = 1
    variable = sgas
  []
  [sbrine]
    type = PorousFlowPropertyAux
    property = saturation
    phase = 0
    variable = sbrine
  []
[]

[Kernels]
  [mass0]
    type = PorousFlowMassTimeDerivative
    variable = pgas
    fluid_component = 0
  []
  [adv0]
    type = PorousFlowAdvectiveFlux
    variable = pgas
    fluid_component = 0
  []
  [mass1]
    type = PorousFlowMassTimeDerivative
    variable = zi
    fluid_component = 1
  []
  [adv1]
    type = PorousFlowAdvectiveFlux
    variable = zi
    fluid_component = 1
  []
  [energy]
    type = PorousFlowEnergyTimeDerivative
    variable = temperature
  []
  [heat_adv]
    type = PorousFlowHeatAdvection
    variable = temperature
  []  
  [conduction]
    type = PorousFlowHeatConduction
    variable = temperature
  []
[]


[Materials]
  [temperature]
    type = PorousFlowTemperature
    temperature = temperature
  []
  [brineCo2Properties]
    type = PorousFlowFluidState
    gas_porepressure = pgas
    z = zi
    temperature = temperature
    xnacl = Xnacl
    capillary_pressure = pc
    fluid_state = fs
  []
  [porosity]
    type = PorousFlowPorosityConst
    porosity = 0.05
  []
  [permeability]
    type = PorousFlowPermeabilityConst
    permeability = '50e-15 0 0   0 50e-15 0   0 0 50e-15'
  []
  [relperm0]
    type = PorousFlowRelativePermeabilityCorey
    n = 4
    s_res = 0.30
    sum_s_res = 0.35
    phase = 0
  []
  [relperm1]
    type = PorousFlowRelativePermeabilityCorey
    n = 4
    s_res = 0.30
    sum_s_res = 0.35
    phase = 1
  []
  [rock_heat]
    type = PorousFlowMatrixInternalEnergy
    specific_heat_capacity = 1000
    density = 2600
  [] 
  [rock_thermal_conductivity]
    type = PorousFlowThermalConductivityIdeal
    dry_thermal_conductivity = '2 0 0  0 2 0  0 0 2'
  []
[]


[DiracKernels]
  [fluid_produce]   
    type = PorousFlowSquarePulsePointSource
    point = '5 -250 0'
    mass_flux = -1.3e-4   # -65  #-0.033   # 
    variable = pgas
  []
[]

[VectorPostprocessors]
  [vars]
    type = NodalValueSampler
    sort_by = x
    variable = 'pgas zi Xnacl'
    execute_on = 'timestep_end'
    outputs = spatial
  []
  [auxvars]
    type = ElementValueSampler
    sort_by = x
    variable = 'sbrine sgas'
    execute_on = 'timestep_end'
    outputs = spatial
  []
[]

[Postprocessors]
  [Pgas]
    type = PointValue
   point =  '5 0 0'
    variable = pgas
  []
  [sgas]
    type = PointValue
    point =  '5 0 0'
    variable = sgas
  []
  [sbrine]
    type = PointValue
    point =  '5 0 0'
    variable = sbrine
  []
  [Xnacl]
    type = PointValue
    point =  '5 0 0'
    variable = Xnacl
  []
[]


[Preconditioning]
  [smp]
    type = SMP
    full = true
    petsc_options_iname = '-ksp_type -pc_type -sub_pc_type -sub_pc_factor_shift_type'
    petsc_options_value = 'gmres bjacobi lu NONZERO'
  []
[]


[Executioner]
  type = Transient
  solve_type = NEWTON
  end_time = 2e6
  nl_max_its = 25
  l_max_its = 100
  dtmax = 1e5
  nl_abs_tol = 1e-20
  [TimeStepper]
    type = IterationAdaptiveDT
    dt = 100
  []
[]

[Outputs]
  exodus = true
  sync_times = '1e4 1e5 2e6'
  [time]
    type = CSV
  []
  [spatial]
    type = CSV
    sync_only = true
  []
[]
```

**Comment by cpgr** [comment, marked as answer]
Yeah, WaterNCG is probably closest but doesn't do brine so no salt. BrineCO2 is functionally equivalent to TOUGHS ECO2N, so not quite what you are after. 

I think extending WaterNCG to be BrineNCG is probably what you want, but it would take some coding. I'm happy to chat about this and work out how to do it.

**Reply by batodon** [reply]
Hello @cpgr, thanks a lot! I’m motivated to know about your willingness to guide me on this. As you might already know, I updated the ticket after thoroughly reviewing the documentation on `PorousFlowBrineCO2` module. It turns out that `PorousFlowBrineCO2` allows high temperatures up to 300℃ (just what I need) compared to TOUGH’S ECO2N. Since CO2 occupies most of NCG (approx.. 90%) and `PorousFlowBrineCO2` already has three components compared to `PorousFlowWaterNCG`, I already began updating the former to include the effect of halite dissolution and precipitation. I will consider switching to the latter. 
My biggest challenge, according to section 2.1 of Blanco et al. (https://doi.org/10.1007/s11242-018-1057-7 )  and section 3.2 of Battistelli et al. (https://www.rockware.com/downloads/documentation/petrasim/Battistelli_Stanford_2017.pdf), is to add an extra phase to represent the solid (halite) phase and accordingly implement the thermodynamic properties of this phase. What do you think? 
So far, I’m implementing the mass fraction of halite in the gas and aqueous phases based on correlations of Driesner and Heinrich (2007) and Palliser and McKibbin (1998). Please, let me know if I’m on the right path. Thanks again!

**Reply by cpgr** [reply]
Before you do too much check out what is in `BrineFluidProperties` - it most likely has the Driesner work already!

**Reply by batodon** [reply]
Thanks for the heads up, @cpgr. You save me from doing extra work. To compute the mass fraction, I need the solubility of the halite in water. `BrineFluidProperties` indeed has this solubility but it is based on Potter et al. (1997). I will update it with Driesner.

**Reply by batodon** [reply]
Hello @cpgr, I updated the `BrineFluidProperties` with the halite solubilities. I called it `BrineFluidPropertiesBeta`. The main modifications are the functions `haliteSolubilityWater` and `haliteSolubilityGas` on lines 123 and 132 in the header file (see attached). It compiles successfully, but I keep getting the following object error when I test it. Any hints, please?

```
*** ERROR ***
A 'FluidProperties' object already exists with the name 'brine:water'.
```

[BrineFluidPropertiesBeta_C.txt](https://github.com/idaholab/moose/files/10994116/BrineFluidPropertiesBeta_C.txt)
[BrineFluidPropertiesBeta_h.txt](https://github.com/idaholab/moose/files/10994117/BrineFluidPropertiesBeta_h.txt)

**Reply by GiudGiud** [reply]
maybe you are unlucky and the name already does exist from the base class?
can you try changing this a little like this
```
  const std::string water_name = name() + ":water_new_new";
  {
    const std::string class_name = "Water97FluidProperties";
    InputParameters params = _app.getFactory().getValidParams(class_name);
    if (_tid == 0)
      _fe_problem.addUserObject(class_name, water_name, params);
```

**Reply by cpgr** [reply]
I think @GiudGiud  is correct, although you might not need to add the water user object again if you are inheriting from BrineFluidProperties? Or else you might have two of them?

**Reply by batodon** [reply]
Hello @GiudGiud  and @cpgr, sorry, I’m probably missing something (maybe in my input file). I made the modifications (to the source file) that @GiudGiud  correctly pointed out and also changed the brine `fluidName` from` brine` to `brine_new`, but I’m getting a new error:

```
*** ERROR ***
/home/iyevugah/MOOSE/projects/ergs/problemEOS/ewasg2.i:66: (UserObjects/fs/brine_fp):
    A valid Brine FluidProperties UserObject must be provided
```
I modified part of the input file above as follows:

```
[FluidProperties]
  [brine_new]
    type = BrineFluidPropertiesBeta
  []
  [co2]
    type = CO2FluidProperties
  []
  [water_new]
    type = Water97FluidProperties
  []
  [tabulated_water]
    type = TabulatedBicubicFluidProperties
    fp = water_new
    temperature_max = 900
    temperature_min = 250
    pressure_max = 1E8
    interpolated_properties = 'density viscosity enthalpy internal_energy'
    fluid_property_file = water97_tabulated_11.csv
  []
  [true_co2]
    type = CO2FluidProperties
  []
  [tabulated_co2]
    type = TabulatedBicubicFluidProperties
    fp = true_co2
    temperature_max = 900
    temperature_min = 250
    pressure_max = 1E8
    interpolated_properties = 'density viscosity enthalpy internal_energy'
    fluid_property_file = co2_tabulated_11.csv
  []
[]

[UserObjects]
  [dictator]
    type = PorousFlowDictator
    porous_flow_vars = 'pgas zi temperature'
    number_fluid_phases = 2
    number_fluid_components = 3
  []
  [pc]
    type = PorousFlowCapillaryPressureConst
    pc = 0
  []
  [fs]
    type = PorousFlowBrineSaltCO2
    brine_fp = brine_new
    water_fp = water_new
    co2_fp = co2          
    capillary_pressure = pc
  []
[]
```

**Reply by cpgr** [reply]
Your PorousFlowBrineSaltCO2 user object will have the following (if it inherits from BrineCO2)

```
  if (_brine_fp.fluidName() != "brine")
    paramError("brine_fp", "A valid Brine FluidProperties UserObject must be provided");
```

You should leave `BrineFluidPropertiesBeta::fluidName() = 'brine'` to avoid this.

**Reply by batodon** [reply]
Awesome! Finally, It works! I think I was overthinking this issue :). Now I can move on and implement the mass fractions and the solid phase thermodynamic properties in my `PorousFlowBrineSaltCO2`. Thanks, @GiudGiud and @cpgr. Please, let me know if you have any feedback on the correlation. I specified the pages and the references in the source file for the `BrineFluidPropertiesBeta`.

**Reply by batodon** [reply]
Hello @GiudGiud  and @cpgr. All is looking good. I successfully implemented the mass fractions of halite and its thermodynamic properties. I, however, ran into a major issue. Since my model includes a solid phase, I must consider a multiphase system (Liquid-gas-solid). The phase state enum from the `PorousFlowFluidStateBase` currently supports only two-phase systems (liquid-gas). Is there a way around it so I can implement a multiphase system? I’m looking at six combinations from a liquid-gas-solid phase instead of the three existing ones from the liquid-gas option (i.e., Liquid, Gas, Twophase). Thanks.

**Reply by cpgr** [reply]
I've been thinking about this and am not sure that you need to change this bit. In my mind, I can see water evaporating into the gas phase (so two phases) until the salt exceeds the solubility limit, in which case it forms a solid. However, that enum is for the gas component which still exists in two phases. I think that you could just have an extra phase with an equivalent saturation, and use the salt mass fraction to populate that phase (and give it zero relative permeability so that it is immobile).

The next phase would be to modify the porosity accordingly, so that it depends on the amount of precipitate.

**Reply by batodon** [reply]
Hello @cpgr. That’s right. I don’t need an extra bit (for the extra phase combinations) since the solid phase is (considered) immobile. I implemented this idea, and my fluid-state object is looking better. I, however, need a few confirmations.

>  "I think that you could just have an extra phase with an equivalent saturation ..."

I initialized the solid phase saturation with a small non-zero value. Is this the way to go, or is there a saturation equation for the solid phase in multiphase flow that I need to implement? I don’t see any reliable equation in this regard. 

> "The next phase would be to modify the porosity accordingly, so that it depends on the amount of precipitate..."

I assume I need to modify the porosity material object used in my input file to account for the solid phase saturation. If this is true, how do I link the solid saturation from my fluid-state object to the porosity material object? Sorry, I’m struggling a bit to grasp this. Any ideas that could get me going? Thanks.

**Reply by cpgr** [reply]
My thought was to have a third phase in the user object - so liquid, gas, and immobile - and give the third phase a zero relative permeability. So in an element there might be liquid (Sl), gas (Sg)  and immobile (Ss) phases, where the 'saturation' of the solid phase is just equal to the fraction of pore space that it fills. I think you could just initialise it at zero (assuming that all the salt is initially dissolved in water).

I would leave modifying the porosity until after you can run a model with no porosity modification. When you are ready, `PorousFlowPorosity` can deal with precipitation, see https://mooseframework.inl.gov/source/materials/PorousFlowPorosity.html.

**Reply by batodon** [reply]
Hello, @cpgr, so I finally run the user object after making the necessary corrections, and it converged. I also modified the porosity to include the effect of the solid phase saturation. I attached a snippet of the results here. These results do not quite correspond well with the results of TOUGH2’s EWASG example. 
Could it be due to how I computed the porosity and the solid saturation? For the porosity, I modified and changed  the `PorousFlowPorosityConst` to `PorousFlowEffectivePorosityConst` as follows:  `phi_{eff} = phi(1 - Ss) `. `Ss` is the solid saturation and `phi` is the total porosity. 

![sat](https://user-images.githubusercontent.com/113298585/229375901-8fbe9d4e-d121-4598-8f36-991b69757132.PNG)
![2](https://user-images.githubusercontent.com/113298585/229375916-f61d4d19-9320-49a1-b8c1-3bafaf039406.PNG)

**Reply by batodon** [reply]
Hello @cpgr, I’m not getting any luck with the ewasg example, so I’m approaching my simulation differently based on the publication by Grimm et al. https://doi.org/10.1029/2020WR027213. Specifically, I’m updating my permeability and porosity models to reflect the salt precipitation when Xnacl reaches (and exceeds) the solubility limit instead of introducing a new (solid) phase. 

Following this approach, the porosity and permeability should evolve as a function of the fracture aperture during halite precipitation. However, I’m unsure about the routine used (in the publication) to code the precipitation rate (`R`) and the fracture aperture (`b`) evolution. Since salt precipitates when the concentration (or the mass fraction in the aqueous phase, i.e., `Xnacl`) reaches (and exceeds) the solubility limit, I’m guessing that `R` and `b` are (coded as)  material property in the `BrineCO2` fluid property user-Object. After that, they were exported as aux-variables, with the fracture aperture eventually used to update the porosity, permeability and capillary pressure. Is this correct, or I’m I missing something? I searched for the repo corresponding to these publications but found nothing useful. I’ll be deeply grateful if you could share any repo regarding this problem that could help me. Thank you.

**Reply by cpgr** [reply]
I can't look at this today, but will do it first thing next week. I want this to work so will allocate some time next week to helping getting this going

**Reply by batodon** [reply]
Thanks for the feedback, @cpgr, It’s very refreshing! I’ll keep working on it too. Hopefully, it works out.

**Reply by batodon** [reply]
Hello, @cpgr. Please, any updates?

**Reply by robo-boluo** [reply]
Hello, @batodon  and @cpgr,

I am also curious about the evolution of porosity and permeability due to salt precipitation and would like to learn of any updates on this topic.
